cdap-ui/app/cdap/components/AbstractWidget/SchemaEditor/index.tsx

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright © 2020 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+import * as React from 'react';
+import withStyles, { WithStyles, StyleRules } from '@material-ui/styles/withStyles';
+import ThemeWrapper from 'components/ThemeWrapper';
+import {
+  SchemaManager,
+  INode,
+  ISchemaManager,
+  IOnChangeReturnType,
+} from 'components/AbstractWidget/SchemaEditor/Context/SchemaManager';
+import { ISchemaType } from 'components/AbstractWidget/SchemaEditor/SchemaTypes';
+import {
+  IFlattenRowType,
+  IFieldIdentifier,
+  IOnChangePayload,
+} from 'components/AbstractWidget/SchemaEditor/EditorTypes';
+import { FieldsList } from 'components/AbstractWidget/SchemaEditor/FieldsList';
+import {
+  SchemaValidatorConsumer,
+  SchemaValidatorProvider,
+} from 'components/AbstractWidget/SchemaEditor/SchemaValidator';
+import { dumbestClone as cloneDeep } from 'services/helpers';
+
+const styles = (): StyleRules => {
+  return {
+    schemaContainer: {
+      width: '100%',
+      height: '100%',
+    },
+  };
+};
+
+interface ISchemaEditorProps extends WithStyles<typeof styles> {
+  schema: ISchemaType;
+  onChange: (props: {
+    tree: INode;
+    flat: IFlattenRowType[];
+    avroSchema: ISchemaType;
+  }) => IOnChangeReturnType;
+}
+
+interface ISchemaEditorState {
+  tree: INode;
+  flat: IFlattenRowType[];
+}
+
+class SchemaEditorBase extends React.Component<ISchemaEditorProps, ISchemaEditorState> {
+  private schema: ISchemaManager = null;
+  constructor(props) {
+    super(props);
+    const { options } = props;
+    this.schema = SchemaManager(this.props.schema, options).getInstance();
+    this.state = {
+      flat: cloneDeep(this.schema.getFlatSchema()),
+      tree: cloneDeep(this.schema.getSchemaTree()),
+    };
+  }
+
+  public componentWillReceiveProps(nextProps) {
```

Comment:
I think for this particular case, we can use the `getDerivedStateFromProps` lifecycle and memoizing it since you need to update the state based on props change. Might future-proof this since componentWillReceiveProps is being deprecated

Reply:
In a future PR I am getting rid of this function. Any component that needs to update the schemaeditor with new schema has to reload the component. It became a hassle when integrating with the plugins.